### PR TITLE
Store encrypted fields list to _ct._cf #2

### PR DESCRIPTION
**Pull request by jenyatra**
### TL;DR

Remember the time when you needed to decrypt an encrypted field? Or maybe you needed to move a field to encryption? 
Remember the pain?

1) Run once:
```
const updateCtAll = async () => {
    for await (const doc of Model.find()) {
        doc.updateCt()
        await doc.save()
    }
}
```

2) Set `cfMode: 'require'` option in plugin configuration

And now you can change schema and plugin options to tweak list of encrypted fields.
Existing documents will work transparently, because each document now includes its own list of encrypted fields!

Keep in mind, that existing documents keep their lists of encrypted fields. So, updating a document won't automaticaly "reconfigure" its ancryption. But you can call `doc.updateCt()` and document will be reconfigured!

So, if you need to add an index over previously encrypted field (or just decrypt a field), or to encrypt another field, you just do it in schema's config. And when you need to reconfigure existing documents, just run something like `updateCtAll` again.

See more info in new Readme sections: https://github.com/jenyatra/mongoose-encryption/tree/cf#changing-options
___________________________________________

Better to analyze this PR by commits: https://github.com/joegoldbeck/mongoose-encryption/pull/109/commits,
commentary provided.
___________________________________________

### PR goals:
1. Protect from data loss caused by changes in encrypted fields list 
2. Provide a way to change encrypted fields list in controlled manner

### Data loss example:
Data loss occurs when schema is changed in a way that affects encrypted fields list, such as adding an index on a field:

```
const Schema1 = new Schema({
    my_data: { type: String },
})

const Schema2 = new Schema({
    my_data: {  type: String, index: true }
})

// Adding index to the Schema2 will lead to exclusion of `my_data` field from `encryptedFields` list.
// Because of that, `my_data` field won't be populated from `_ct` on next instantiation,
// Saving document with non-populated `my_data` field will overwrite `_ct`, and data will be lost.

// Same problem occurs when fields excluded from encryption other ways,
// such as using `excludeFromEncryption` option.
```

### Tests repo:
More detailed examples (ready to run): https://github.com/jenyatra/mongoose-encryption-test
```
git clone https://github.com/jenyatra/mongoose-encryption-test
cd mongoose-encryption-test
npm install
echo "MONGO_URI=mongodb://localhost:27017/$SET YOUR TEST DB HERE$" > .env
```
Available tests:

- `npm run fields_change`: Test changes in encrypted fields list for root fields
- `npm run fields_change:cf`: Previous test run by proposed version
- `npm run fields_change_deep`:  Same for nested fields (type: Mixed, dot notation)
- `npm run fields_change_deep:cf`:  Previous test run by proposed version
- `npm run fields_change_nested`: Test for nested docs (type: new Schema)
- `npm run fields_change_nested:cf`: Previous test run by proposed version
- `npm run init_cf`: Example of _cf initialization on existing data

### Addressing issues on changing encryption fields list
Currently there's no way to easily decrypt an encrypted field, or vice versa. Just adding a field to `excludeFromEncryption` won't work, and could lead to a data loss.

Related issues:
- https://github.com/joegoldbeck/mongoose-encryption/issues/83
- https://github.com/joegoldbeck/mongoose-encryption/issues/48

Using new features proposed in this PR, Schema and plugin options related to encryption fields list can be changed seamlessly.

### How it works
Encrypted fields list stored to `_cf` property of secret part (`_ct._cf`).

On decrypt, fields listed in `_cf` will be assigned to data, additionally `_cf` list is assigned to `Document` instance, to use in following encryption.

Encrypt uses fields list from `_cf` property of `Document` instance if it's set, otherwise current schema's list is used.

On executing `updateCf`, fields that are no longer in current schema's list are marked modified (to force its update by mongoose). Current schema's list is assigned to `_cf` property of `Document` instance (to force new encryption rules).

New `cfMode` option is added:

- `disable` (default): Disable new function
- `require`: requires `_cf` list to operate on existing documents
- `maintenance`: uses `_cf` list when set on a document, otherwise uses Schema's `encryptedFields`

### Backward compatibility and default behavior
PR keeps backward compatibility in regular cases.
New features are disabled by default.

### CODE REVIEW

File: lib/plugins/mongoose-encryption.js [hunk 1]
```diff
@@ -360,11 +360,10 @@ var mongooseEncryption = function(schema, options) {
         cb();
     };
 
-    schema.methods.decryptSync = function() {
```
Comment (jenyatra) [inline lib/plugins/mongoose-encryption.js:363]: Initially, `decryptSync` has no parameters, and were run in execution context of the object to be decrypted. Usually it's `Document` itself, but in `init` hook we need to decrypt `data` object.

We'll need access to Document later for `cf` support.

To keep backward compatibility, `decryptData` method extracted from `decryptSync`, which allows to decrypt object provided in first parameter, running in execution context of `Document`.

File: test/encrypt.coffee [hunk 2]
```diff
@@ -398,7 +398,7 @@ describe 'EncryptedModel.find()', ->
   before (done) ->
     @sandbox = sinon.sandbox.create()
     @sandbox.spy BasicEncryptedModel.prototype, 'authenticateSync'
-    @sandbox.spy BasicEncryptedModel.prototype, 'decryptSync'
```
Comment (jenyatra) [inline test/encrypt.coffee:401]: Now using `decryptData` in `init` hook

File: lib/plugins/mongoose-encryption.js (hunk 1 above)
Comment (jenyatra) [reply lib/plugins/mongoose-encryption.js:363]: Basically `decryptData` is same as `decryptSync`, but `this` changed to `data` parameter.

File: lib/plugins/mongoose-encryption.js [hunk 3]
```diff
@@ -321,12 +341,22 @@ var mongooseEncryption = function(schema, options) {
...
+                cf = encryptedFields;
+            } else {
+                cf = that._cf || encryptedFields;
+            }
+
             cipher = crypto.createCipheriv(ENCRYPTION_ALGORITHM, encryptionKey, iv);
-            objectToEncrypt = pick(that, encryptedFields, {excludeUndefinedValues: true});
+            objectToEncrypt = pick(that, cf, {excludeUndefinedValues: true});
+            if (cfMode !== 'disable') {
+                objectToEncrypt._cf = cf.join(',');
```
Comment (jenyatra) [inline lib/plugins/mongoose-encryption.js:358]: Here `_cf` list is stored

File: lib/plugins/mongoose-encryption.js [hunk 4]
```diff
@@ -343,14 +373,44 @@ var mongooseEncryption = function(schema, options) {
             that._ct = Buffer.concat([VERSION_BUF, iv, Buffer.from(encrypted, 'utf8')]);
 
             // remove encrypted fields from cleartext
-            encryptedFields.forEach(function(field){
-                setFieldValue(that, field, undefined);
+            cf.forEach(function(field){
+                that.set(field, undefined);
```
Comment (jenyatra) [inline lib/plugins/mongoose-encryption.js:377]: This change needed to ensure nested fields marked modified.

File: lib/plugins/mongoose-encryption.js [hunk 5]
```diff
@@ -380,7 +440,17 @@ var mongooseEncryption = function(schema, options) {
                 throw new Error('Error parsing JSON during decrypt of ' + idString + ': ' + err);
             }
 
-            encryptedFields.forEach(function(field) {
+            if (cfMode === 'require' && !decryptedObject._cf) {
+                throw new Error('Cipher fields list _cf is required, but not found in cipher text');
+            } else if (cfMode !== 'disable' &&  decryptedObject._cf) {
+                cf = decryptedObject._cf.split(',');
```
Comment (jenyatra) [inline lib/plugins/mongoose-encryption.js:455]: Here `_cf` list is read

File: lib/plugins/mongoose-encryption.js [hunk 6]
```diff
@@ -394,6 +464,10 @@ var mongooseEncryption = function(schema, options) {
 
             data._ct = undefined;
             data._ac = undefined;
+
+            if (cfMode !== 'disable') {
+                this._cf = cf;
```
Comment (jenyatra) [inline lib/plugins/mongoose-encryption.js:478]: Assigning cf to Document instance, so it will remember its configuration.

File: lib/plugins/mongoose-encryption.js [hunk 7]
```diff
@@ -264,6 +280,10 @@ var mongooseEncryption = function(schema, options) {
         schema.pre('save', function(next) {
             var that = this;
             if (this.isNew || this.isSelected('_ct') ){
+                if (cfMode !== 'disable' && this.isNew && !this._cf) {
+                    this._cf = encryptedFields;
```
Comment (jenyatra) [inline lib/plugins/mongoose-encryption.js:284]: Assigning Schema's default cf for new documents

File: lib/plugins/mongoose-encryption.js [hunk 8]
```diff
@@ -321,12 +341,22 @@ var mongooseEncryption = function(schema, options) {
...
-            var cipher, jsonToEncrypt, objectToEncrypt;
+            var cf, cipher, jsonToEncrypt, objectToEncrypt;
             if (err) {
                 return cb(err);
             }
+
+            if (cfMode === 'disable') {
+                cf = encryptedFields;
+            } else {
+                cf = that._cf || encryptedFields;
```
Comment (jenyatra) [inline lib/plugins/mongoose-encryption.js:352]: When Document instance have helper _cf propety, it defines the list of encrypted fields.

File: lib/plugins/mongoose-encryption.js [hunk 9]
```diff
@@ -420,17 +420,22 @@ var mongooseEncryption = function(schema, options) {
         cb();
     };
 
+    schema.methods.decryptCt = function(_ct) {
```
Comment (jenyatra) [inline lib/plugins/mongoose-encryption.js:427]: Ever wonder what's inside the `_ct`?
Now we can get it.

**Comment by joegoldbeck** [comment]
Hi @jenyatra – appreciate this PR and the thought and effort you've been putting into this issue. And apologies for my delay in feedback!

I was looking through your solution, and it poked me to take a deeper look myself, because my suspicion is that this can be solved without adding additional options or stored data. What I'm thinking is, when decrypting `_ct` we know what fields are there by inspection, perhaps we don't actually need to store those fields separately. The existing bad behavior, I think, comes from [here](https://github.com/joegoldbeck/mongoose-encryption/blob/803fc3c8743586debb7e94cb00ae8aa4ee3c50a0/lib/plugins/mongoose-encryption.js#L384) where the decrypted fields that get added to the returned record are only those listed in `encryptedFields`, but I'm not sure there's a reason to only add those fields.

Following that thread for a bit, I created #110.

What I'm not sure of is whether there's some downside to this approach. There may have been some reason (other than convenience) that past-self did it the way it is now, but I'm currently drawing a blank 😅.

No pressure to put any more time into this, but if you're up for it, I'd love to run a couple questions by you:

Having sat with the code for a while, have you thought of situations where we really need to track the encrypted field options explicitly and _only_ decrypt those fields, or does #110 cover it?

I see you have a handful of test cases in another repo which is awesome ( though fwiw, best to include those as part of a PR, adding to the unit tests in the repo). Are there edge cases that I'm not covering in the current test that you've thought of via your own PR that may be worth adding?

Thanks for taking the time you've already taken to dig into this!

**Comment by jenyatra** [comment]
Hi!
You solution looks much better than mine. For certain, I've overcomplicated this (but nothing can beat my first PR, that was a real monstrosity :)

I think there might be a little problem (easily fixable): when add encryption for a nested field, values can be duplicated both in clear and cipher text upon existing record update. It can be fixed like [this](https://github.com/joegoldbeck/mongoose-encryption/pull/109/files#r780766458). I'll check #110 for this issue in couple hours.

**Comment by jenyatra** [comment]
I think it may be possible to combine this PR and #110.
Currently, when `_cf` is not found, it either throws error (when `cfMode: ''require`), or uses `encryptedFields` as a default (when `cfMode: 'maintenance'`).

Maybe it can use #110 essence as default:
```
var fieldsToDecrypt = objectUtil.getPaths(decryptedObject)
```
This way, the complicated `_cf` initialization procedure is not required, which means users can change schemas right away (as simple as intended in #110)

I'll check this possibility if #110 won't work.
